pkg/workflow/workflow.go

```diff
@@ -227,6 +228,64 @@ func (w *workflow) GetBackoffWaitTime() time.Duration {
 	return time.Second
 }
 
+func (w *workflow) HandleSuspendWait(ctx monitorContext.Context) (doWaiting bool, durationWaiting time.Duration, errRet error) {
+	ctx.Info("handle suspend wait")
+	for i, stepStatus := range w.app.Status.Workflow.Steps {
+		if !w.isWaitSuspendStep(stepStatus) {
+			continue
+		}
+
+		step := w.getWorkflowStepByName(stepStatus.Name)
+		if step.Name == "" {
+			errRet = fmt.Errorf("failed to get workflow step by name: %s", stepStatus.Name)
+			return
+		}
+
+		d, wd, err := wfTasks.GetSuspendStepDurationWaiting(step)
+		if err != nil {
+			ctx.Error(err, "failed to get suspend step wait duration")
+			errRet = err
+			return
+		}
+
+		if d {
+			doWaiting = d
+			durationReserve := wd - time.Since(stepStatus.FirstExecuteTime.Time)
+			if durationReserve <= 0 {
+				w.app.Status.Workflow.Steps[i].Phase = common.WorkflowStepPhaseSucceeded
+			}
+
+			if durationReserve > 0 && (durationWaiting > durationReserve || durationWaiting <= 0) {
+				suspendState := fmt.Sprintf("durationWaiting(%s)", wd.String())
+				if w.app.Status.Workflow.SuspendState != suspendState {
+					w.app.Status.Workflow.SuspendState = suspendState
+				}
+				durationWaiting = durationReserve
+			}
+		}
+
+		if !w.dagMode {
+			return
+		}
+	}
+
+	return doWaiting, durationWaiting, errRet
+}
+
+func (w *workflow) isWaitSuspendStep(status common.WorkflowStepStatus) bool {
+	return status.Type == wfTypes.WorkflowStepTypeSuspend && status.Phase == common.WorkflowStepPhaseRunning
```

Comment:
Can we use `strings.Contains(w.status.suspendState, durationWaiting)` here?
Also, I think it's better if we set `durationWaiting` to a const.

Reply:
we include checking the suspendState here.

Reply:
oh, i see, this is a step level status